### PR TITLE
itest: continued itest refactor and fix - I

### PR DESCRIPTION
**Pull request by yyforyongyu**
Depending on #6759, this PR continues the refactor work to fix the itest flakes. A few bugs have been found and documented in new issues along the way. Most of the work has been dedicated to making `icase=test_multi-hop_htlc` to pass on all backends. I think due to the current itest situation, this test failed or even didn't get a chance to be run due to its preceding tests failed.

These might be the only two "hard" PRs. The rest would be straightforward, just replacing old patterns and fixing flakes if found.

Only the last 28 commits are relevant.

Depends on #6868

### CODE REVIEW

File: lntest/itest/list_on_test.go [hunk 1]
```diff
@@ -16,8 +16,32 @@ var allTestCasesTemp = []*lntemp.TestCase{
 		TestFunc: testBasicChannelFunding,
 	},
 	{
-		Name:     "test multi-hop htlc",
-		TestFunc: testMultiHopHtlcClaims,
+		Name:     "multi hop htlc local timeout",
```
Comment (guggero) [inline lntest/itest/list_on_test.go:19]: Ah nice, this will spread the duration of the individual tests much better across the parallel running tranches! With all your improvements we might even be able to increase the parallelism to 6 or 8 concurrent tests (need to experiment with that).
So smaller/shorter tests are way better :tada:

**Review by guggero — COMMENTED** [review]
Wow, this has to be among the most meticulous work I've seen in a very long time :astonished: :rocket: :clap: 
Awesome commit structure and amazing refactor work all around.
Also love the split of the tests into smaller parts, that way we might be able to get the parallel running tests balanced out better.

File: lntemp/harness.go [hunk 2]
```diff
@@ -676,3 +674,78 @@ func (h *HarnessTest) CloseChannel(hn *node.HarnessNode,
...
+		require.Equal(h, pkScriptStr, expPkScriptStr,
+			"pkscript mismatch")
+	}
+
+	// If the transaction should remain unconfirmed, then we'll wait until
+	// the target node's unconfirmed balance reflects the expected balance
+	// and exit.
+	if !confirmed && !h.IsNeutrinoBackend() {
+		expectedBalance := btcutil.Amount(
+			initialBalance.UnconfirmedBalance) + amt
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/harness.go]: fixed.

File: lntemp/harness_assertion.go [hunk 3]
```diff
@@ -498,3 +502,134 @@ func (h *HarnessTest) WaitForGraphSync(hn *node.HarnessNode) {
...
+	max, min int32) []*lnrpc.Utxo {
+
+	old := hn.State.UTXO.Confirmed
+	if max == 0 {
+		old = hn.State.UTXO.Unconfirmed
+	}
+
+	var utxos []*lnrpc.Utxo
+	err := wait.NoError(func() error {
+		req := &lnrpc.ListUnspentRequest{
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/harness_assertion.go]: Nope it just was there and I didn't know the newer one🤦🏻 Switched to the newer one.

File: lntemp/harness_assertion.go [hunk 4]
```diff
@@ -498,3 +502,134 @@ func (h *HarnessTest) WaitForGraphSync(hn *node.HarnessNode) {
 	}, DefaultTimeout)
 	require.NoError(h, err, "%s: timeout while sync to graph", hn.Name())
 }
+
+// AssertNumUTXOs waits for the given number of UTXOs to be available or fails
+// if that isn't the case before the default timeout.
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/harness_assertion.go]: Def! Comments added.

File: lntest/itest/lnd_funding_test.go [hunk 5]
```diff
@@ -133,15 +132,10 @@ test:
...
-				"---- basic channel funding subtest %s ----\n",
-				testName,
-			)
-			net.Alice.AddToLogf(logLine)
-
-			success := t.t.Run(testName, func(t *testing.T) {
-				testFunding(cc, dc)
+			success := ht.Run(testName, func(t *testing.T) {
+				st, cleanup := ht.Subtest(t)
+				defer cleanup()
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_funding_test.go]: Yeah actually was gonna add it in the last PR as I realized its existence. But I guess it's better to have it here, so added.

File: lntest/itest/lnd_funding_test.go [hunk 6]
```diff
@@ -224,21 +210,17 @@ func basicChannelFundingTest(t *harnessTest, net *lntest.NetworkHarness,
...
-	require.NoError(t.t, err, "unable to obtain chan")
+	aliceChannel := ht.GetChannelByChanPoint(alice, chanPoint)
+	bobChannel := ht.GetChannelByChanPoint(bob, chanPoint)
 
 	closeChan := func() {
 		// Finally, immediately close the channel. This function will
 		// also block until the channel is closed and will additionally
 		// assert the relevant channel closing post conditions.
-		closeChannelAndAssert(t, net, alice, chanPoint, false)
+		ht.CloseChannel(alice, chanPoint, false)
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_funding_test.go]: Done.

File: lntest/itest/lnd_funding_test.go [hunk 7]
```diff
@@ -533,16 +529,25 @@ func testExternalFundingChanPoint(ht *lntemp.HarnessTest) {
...
 	// HTLCs.
 	ht.AssertInvoiceSettled(dave, resp.PaymentAddr)
 
+	// TODO(yy): remove the sleep once the following bug is fixed.
+	// When the invoice is reported settled, the commitment dance is not
+	// yet finished, which can cause an error when closing the channel,
+	// saying there's active HTLCs. We need to investigate this issue and
+	// reverse the order to, first finish the commitment dance, then report
+	// the invoice as settled.
+	time.Sleep(2 * time.Second)
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_funding_test.go:543]: Actually I found this issue after refactoring the `CloseChannel` by removing the `wait.NoError`. Ever since then I suspect the `wait.NoError` is hiding something. Later on, in following PRs, this error popped up multiple times in other tests, which annoyed me a bit, leading me to think maybe I should put it in the `CloseChannelAssertPending`. But I decided to choose not to, and want to mark a TODO whenever I see it fail, mainly due to three reasons,
- other places have a similar error, but a different precondition. In this test it's after the invoice is reported as settled, in other cases, it's after payment is reported as succeeded, or closing channel right after opening. I think they have different causes and want to fix them one by one.
- I want to mark the specific places in the test so it's easier to come back to locate and fix them.
- I intentionally do this ugly and hacky `time.Sleep` style to address what's hidden in a `wait` and more importantly remind myself to fix them asap(the plan is to get itest pass, and start fixing the issues found right after).

File: lntest/itest/list_on_test.go (hunk 1 above)
Comment (yyforyongyu) [reply lntest/itest/list_on_test.go:19]: I tried with 8 locally but it seemed to slow down the test quite a bit, maybe it's my machine not dealing with goroutine schedulers smart enough🧐. Anyway totally agree with smaller/shorter tests are better. I find them not only run faster but also easier to debug.

File: lntest/itest/lnd_funding_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (guggero) [reply lntest/itest/lnd_funding_test.go:543]: Hmm, okay. Sounds good to me, as long as we have an easy pattern to search for :+1:

**Review by Crypt-iQ — COMMENTED** [review]
looks like a flake in TestLightningNetworkDaemonTemp/tranche01/09-of-104/btcd/multi_hop_htlc_aggregation/zeroconf=false/committype=ANCHORS

File: lntemp/harness_assertion.go [hunk 8]
```diff
@@ -682,3 +683,134 @@ func (h *HarnessTest) GetChannelCommitType(hn *node.HarnessNode,
...
+
+	return target
+}
+
+// ReceivePaymentUpdate waits until a message is received on the payment client
+// stream or the timeout is reached.
+func (h *HarnessTest) ReceivePaymentUpdate(
+	stream rpc.PaymentClient) *lnrpc.Payment {
+
+	chanMsg := make(chan *lnrpc.Payment)
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/harness_assertion.go:850]: done

File: lntest/itest/lnd_funding_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_funding_test.go:543]: Cool will remove this `TODO` once these changes are made.

File: lntest/itest/lnd_channel_backup_test.go [hunk 9]
```diff
@@ -1217,54 +1211,18 @@ func testDataLossProtection(net *lntest.NetworkHarness, t *harnessTest) {
...
-		payReqs, _, _, err := createPayReqs(
-			node, paymentAmt, numInvoices,
-		)
-		if err != nil {
-			t.Fatalf("unable to create pay reqs: %v", err)
-		}
-
-		// Wait for Carol to receive the channel edge from the funding
-		// manager.
-		err = carol.WaitForNetworkChannelOpen(chanPoint)
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_channel_backup_test.go]: This method is now renamed to `AssertTopologyChannelOpen` to differentiate it from other channel open assertions, and it's asserted when calling `OpenChannel`. It's refactored this way so other tests don't need to care about this process. Instead, the whole funding flow is tested in funding related tests.

File: lntest/itest/lnd_multi-hop_test.go [hunk 10]
```diff
@@ -95,63 +100,56 @@ func testMultiHopHtlcClaims(net *lntest.NetworkHarness, t *harnessTest) {
 		},
 	}
 
-	for _, typeAndConf := range commitWithZeroConf {
-		typeAndConf := typeAndConf
-		testName := fmt.Sprintf(
-			"committype=%v zeroconf=%v",
-			typeAndConf.commitType.String(), typeAndConf.zeroConf,
-		)
+	runTestCase := func(st *lntemp.HarnessTest, name string, tc testCase,
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_multi-hop_test.go]: Cool good to know. I think eventually we want to use standby nodes as much as possible to run the tests. I think in a follow-up PR I can make the change once I fix splitting up the log files.

File: lntemp/harness_assertion.go [hunk 11]
```diff
@@ -1026,3 +1028,221 @@ func (h *HarnessTest) AssertActiveHtlcs(hn *node.HarnessNode,
...
 	require.NoError(h, err, "timeout checking active HTLCs")
 }
+
+// ReceiveSingleInvoice waits until a message is received on the subscribe
+// single invoice stream or the timeout is reached.
+func (h *HarnessTest) ReceiveSingleInvoice(
+	stream rpc.SingleInvoiceClient) *lnrpc.Invoice {
+
+	chanMsg := make(chan *lnrpc.Invoice)
+	errChan := make(chan error)
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/harness_assertion.go:1061]: done

File: lntemp/rpc/invoices.go [hunk 12]
```diff
@@ -24,3 +24,61 @@ func (h *HarnessRPC) LookupInvoiceV2(
...
+	req := &invoicesrpc.SettleInvoiceMsg{Preimage: preimage}
+	resp, err := h.Invoice.SettleInvoice(ctxt, req)
+	h.NoError(err, "SettleInvoice")
+
+	return resp
+}
+
+// CancelInvoice cancels a given invoice and asserts.
+func (h *HarnessRPC) CancelInvoice(
+	preimage []byte) *invoicesrpc.CancelInvoiceResp {
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/rpc/invoices.go:57]: done

File: lntest/itest/lnd_multi-hop_htlc_local_chain_claim_test.go [hunk 13]
```diff
@@ -101,73 +88,55 @@ func testMultiHopHtlcLocalChainClaim(net *lntest.NetworkHarness, t *harnessTest,
...
 	numBlocks := padCLTV(uint32(invoiceReq.CltvExpiry -
 		lncfg.DefaultIncomingBroadcastDelta))
+	ht.MineBlocksAssertNodesSync(numBlocks)
 
-	_, err = net.Miner.Client.Generate(numBlocks)
-	require.NoError(t.t, err)
-
-	// Carol's commitment transaction should now be in the mempool. If there
-	// is an anchor, Carol will sweep that too.
-	_, err = waitForNTxsInMempool(
```
Comment (yyforyongyu) [inline lntest/itest/lnd_multi-hop_htlc_local_chain_claim_test.go:144]: Carol's anchor sweep.

File: lntest/itest/lnd_multi-hop_htlc_local_chain_claim_test.go [hunk 14]
```diff
@@ -228,52 +194,25 @@ func testMultiHopHtlcLocalChainClaim(net *lntest.NetworkHarness, t *harnessTest,
...
-					"advanced to the first stage but was "+
-					"stage: %v", c.PendingHtlcs[0].Stage)
-			}
-
-			return nil
-		},
-	)
-	require.NoError(t.t, err)
+	ht.AssertNumPendingForceClose(bob, numPendingChans)
+	ht.AssertNumHTLCsAndStage(bob, aliceChanPoint, 1, 1)
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_multi-hop_htlc_local_chain_claim_test.go]: Fixed.

File: lntest/itest/lnd_multi-hop_htlc_aggregation_test.go [hunk 15]
```diff
@@ -306,81 +268,53 @@ func testMultiHopHtlcAggregation(net *lntest.NetworkHarness, t *harnessTest,
...
-					"stage: %v", c.PendingHtlcs[0].Stage)
-			}
-
-			return nil
-		},
-	)
-	require.NoError(t.t, err)
+	// For this channel, we also check the number of HTLCs and the stage
+	// are correct.
+	ht.AssertNumHTLCsAndStage(bob, bobChanPoint, numInvoices*2, 2)
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_multi-hop_htlc_aggregation_test.go]: In the old check, because we exit at this line `if c.Channel.LocalBalance != 0 {`, this pending channel's HTLCs are never checked. And if I got it right, I think the stage is set to 2 here,
https://github.com/lightningnetwork/lnd/blob/b387e2c718e70afdbf8dff69c02aa7a706ca7dec/contractcourt/htlc_timeout_resolver.go#L439-L442

**Comment by Roasbeef** [comment]
Can be rebased now!

File: contractcourt/htlc_outgoing_contest_resolver.go [hunk 16]
```diff
@@ -112,20 +112,18 @@ func (h *htlcOutgoingContestResolver) Resolve() (ContractResolver, error) {
...
 			//
-			// TODO(joostjager): Statement above may not be valid.
-			// For CLTV locks, the expiry value is the last
-			// _invalid_ block. The likely reason that this does not
-			// create a problem, is that utxonursery is checking the
-			// expiry again (in the proper way).
-			//
-			// Source:
-			// https://github.com/btcsuite/btcd/blob/991d32e72fe84d5fbf9c47cd604d793a0cd3a072/blockchain/validate.go#L154
+			// NOTE: when broadcasting this transaction, btcd will
```
Comment (guggero) [inline contractcourt/htlc_outgoing_contest_resolver.go:119]: This almost sounds like it should be in its own PR?

File: lntemp/harness.go [hunk 17]
```diff
@@ -772,6 +797,8 @@ func (h *HarnessTest) OpenChannel(alice, bob *node.HarnessNode,
 // rpc is called, it will consume one event and assert it's a close pending
 // event. In addition, it will check that the closing tx can be found in the
 // mempool.
+//
+// nolint: ireturn
```
Comment (guggero) [inline lntemp/harness.go:801]: Yeah, let's remove that linter instead.

File: lntemp/harness_assertion.go [hunk 18]
```diff
@@ -820,6 +837,7 @@ func (h *HarnessTest) assertPaymentStatusWithTimeout(stream rpc.PaymentClient,
 	timeout time.Duration) *lnrpc.Payment {
 
 	var target *lnrpc.Payment
+
```
Comment (guggero) [inline lntemp/harness_assertion.go:840]: I don't really like this extra line after a var declaration... What linter is that? Maybe we should tweak or disable that one too.

**Review by guggero — COMMENTED** [review]
Very nice!
Got a few minor comments and noticed a diff in one of the move commits. Also the last commit with the linter fixes can probably be smaller after we merge #7076.

File: lntemp/harness.go (hunk 17 above)
Comment (yyforyongyu) [reply lntemp/harness.go:801]: yep removed!

File: lntemp/harness_assertion.go [hunk 19]
```diff
@@ -712,7 +722,8 @@ func (h *HarnessTest) WaitForBalanceUnconfirmed(hn *node.HarnessNode,
 // Random32Bytes generates a random 32 bytes which can be used as a pay hash,
 // preimage, etc.
 func (h *HarnessTest) Random32Bytes() []byte {
-	randBuf := make([]byte, 32)
+	const numBytes = 32
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/harness_assertion.go:725]: Cool switched to `lntypes.HashSize`.

File: lntemp/harness_assertion.go (hunk 18 above)
Comment (yyforyongyu) [reply lntemp/harness_assertion.go:840]: yeah it gave me a headache...it's from `wsl` I think. Somehow I ended up getting it tho, forcing us to group `var`s together and emphasize them. Also looked into this linter and it can be configured to allow `var` followed by statements. Anyway it's disabled atm.

File: lntemp/harness_node_manager.go [hunk 20]
```diff
@@ -206,16 +215,18 @@ func (nm *nodeManager) initWalletAndNode(hn *node.HarnessNode,
 	// Now that the wallet is unlocked, before creating an authed
 	// connection we will close the old unauthed connection.
 	if err := hn.CloseConn(); err != nil {
-		return nil, fmt.Errorf("close unauthed conn failed")
+		return nil, errors.New("close unauthed conn failed")
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/harness_node_manager.go:218]: yep removed.

File: lntemp/node/harness_node.go [hunk 21]
```diff
@@ -766,19 +769,23 @@ func (hn *HarnessNode) BackupDb() error {
 	return nil
 }
 
-// RestoreDb restores a database backup.
-func (hn *HarnessNode) RestoreDb() error {
+// RestoreDB restores a database backup.
+//
+//nolint:nestif
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/node/harness_node.go:774]: configured to allow more nested ifs.

File: lntemp/harness.go [hunk 22]
```diff
@@ -676,3 +674,79 @@ func (h *HarnessTest) CloseChannel(hn *node.HarnessNode,
...
+	addr := h.DecodeAddress(resp.Address)
+	addrScript := h.PayToAddrScript(addr)
+
+	// Generate a transaction which creates an output to the target
+	// pkScript of the desired amount.
+	output := &wire.TxOut{
+		PkScript: addrScript,
+		Value:    int64(amt),
+	}
+	h.Miner.SendOutput(output, 7500)
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/harness.go]: Changed!

File: lntemp/rpc/lnd.go [hunk 23]
```diff
@@ -214,3 +215,27 @@ func (h *HarnessRPC) CloseChannel(
...
+}
+
+// FundingStateStepAssertErr makes a RPC call to FundingStateStep and asserts
+// there's an error.
+func (h *HarnessRPC) FundingStateStepAssertErr(m *lnrpc.FundingTransitionMsg) {
+	ctxt, cancel := context.WithTimeout(h.runCtx, DefaultTimeout)
+	defer cancel()
+
+	_, err := h.LN.FundingStateStep(ctxt, m)
+	require.Error(h, err, "duplicate pending channel ID funding shim "+
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/rpc/lnd.go:239]: Updated.

File: lntemp/harness.go [hunk 24]
```diff
@@ -676,3 +674,79 @@ func (h *HarnessTest) CloseChannel(hn *node.HarnessNode,
...
+// IsNeutrinoBackend returns a bool indicating whether the node is using a
+// neutrino as its backend. This is useful when we want to skip certain tests
+// which cannot be done with a neutrino backend.
+func (h *HarnessTest) IsNeutrinoBackend() bool {
+	return h.manager.chainBackend.Name() == NeutrinoBackendName
+}
+
+// fundCoins attempts to send amt satoshis from the internal mining node to the
+// targeted lightning node. The confirmed boolean indicates whether the
+// transaction that pays to the target should confirm.
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/harness.go:905]: updated!

File: lntemp/harness_assertion.go [hunk 25]
```diff
@@ -895,3 +895,58 @@ func (h *HarnessTest) AssertNodeNumChannels(hn *node.HarnessNode,
...
+		// Check local balance.
+		if target.LocalBalance == balance {
+			return nil
+		}
+
+		return fmt.Errorf("balance is incorrect, got %v, expected %v",
+			target.LocalBalance, balance)
+	}, DefaultTimeout)
+
+	require.NoError(h, err, "timeout while chekcing for balance")
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntemp/harness_assertion.go:975]: fixed!

File: lntest/itest/lnd_multi-hop_test.go [hunk 26]
```diff
@@ -366,3 +369,1607 @@ func createThreeHopNetwork(ht *lntemp.HarnessTest,
...
+	// is no link anymore to remove the htlc from the commitment tx. For
+	// this test, it is important to actually settle and not leave the
+	// invoice in the accepted state, because without a known preimage, the
+	// channel arbitrator won't go to chain.
+	carol.RPC.SettleInvoice(preimage[:])
+
+	// We'll now mine enough blocks so Carol decides that she needs to go
+	// on-chain to claim the HTLC as Bob has been inactive.
+	numBlocks := padCLTV(uint32(invoiceReq.CltvExpiry-
+		lncfg.DefaultIncomingBroadcastDelta) - 1)
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply lntest/itest/lnd_multi-hop_test.go:983]: hawk eyes! Yeah they deserve to be in their own commits since they are actually fixing the current tests. Mostly it's an extra unwanted block that was mined, and they are now fixed. Meanwhile, when breaking the tests, I further emphasize this issue by using a variable `blocksMined` to track how many blocks are mined so far.

**Review by Crypt-iQ — APPROVED** [review]
